File: 077_Combinations/077_1.py
from typing import List


class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res = []
        tmp = []
        def dfs(cur):
            if len(tmp) == k:
                res.append(tmp[:])
                return
            for i in range(cur, n + 1):
                tmp.append(i)
                dfs(i+1)
                tmp.pop()
        dfs(1)
        return res


if __name__ == "__main__":
    s = Solution()
    print(s.combine(4, 2))

File: 077_Combinations/077_1.java
class Solution {
    private List<List<Integer>> res;
    private Deque<Integer> tmp;

    public List<List<Integer>> combine(int n, int k) {
        res = new ArrayList<>();
        tmp = new LinkedList<>();
        dfs(n, k, 1);
        return res;
    }

    public void dfs(int n, int k, int index) {
        if (tmp.size() == k) {
            res.add(new ArrayList(tmp));
            return;
        }

        // for (int i = index; i <= n; i++)
        for (int i = index; i <= n - (k - tmp.size()) + 1; i++) {
            tmp.add(i);
            dfs(n, k, i + 1);
            tmp.removeLast();
        }
    }
}